### PR TITLE
[Migrillian] Verify Trillian tree before fetching

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This PR makes Migrillian's `Controller` query Trillian tree state before putting any entries to it:
- Makes sure the `Tree` is a `PREORDERED_LOG`.
- Makes sure the tree head is correctly signed.
- Makes sure the tree head is consistent with the CT log's STH.
- Resets the `Fetcher` state so that it doesn't fetch entries already stored in Trillian.

The last two items are useful when `Controller` is run on a log which already has some entries, e.g. when a leader for the tree has been re-elected, or Migrillian was restarted after a fault.

Additionally, this PR moves Trillian-specific code to a separate file, places some TODOs and slightly restructures `main.go`.

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
PTAL.

File: trillian/migrillian/core/controller.go [hunk 1]
```diff
@@ -43,27 +46,52 @@ type Options struct {
...
-func NewController(opts Options) *Controller {
+func NewController(opts Options, ctClient *client.LogClient, trClient *TrillianTreeClient) *Controller {
 	bufferSize := opts.Submitters * opts.BatchesPerSubmitter
 	batches := make(chan scanner.EntryBatch, bufferSize)
-	return &Controller{opts, batches}
+	return &Controller{opts: opts, batches: batches, ctClient: ctClient, trClient: trClient}
 }
 
 // Run transfers CT log entries obtained via the CT log client to a Trillian
 // log via the other client. If Options.Continuous is true then the migration
```
Comment (pavelkalinnikov) [inline trillian/migrillian/core/controller.go]: self-nit: Update the comment.

File: trillian/migrillian/core/controller.go [hunk 2]
```diff
@@ -43,27 +46,52 @@ type Options struct {
...
+	root, err := c.trClient.getVerifiedRoot(ctx)
+	if err != nil {
+		return err
+	}
+	if c.opts.Continuous { // Ignore range parameters in Continuous mode.
+		// TODO(pavelkalinnikov): Restore fetching state from storage in a better
+		// way than "take the current tree size".
+		c.opts.StartIndex, c.opts.EndIndex = int64(root.TreeSize), 0
+		glog.Warningf("Migrating tree %d continuously: updated entry range to [%d, %d)",
+			c.trClient.tree.TreeId, c.opts.StartIndex, c.opts.EndIndex)
```
Comment (pavelkalinnikov) [inline trillian/migrillian/core/controller.go]: self-nit: EndIndex is 0, maybe print "infinity" instead?

**Comment by pavelkalinnikov** [comment]
ping

File: trillian/ctfe/handlers.go [hunk 3]
```diff
@@ -723,8 +723,10 @@ func getRoots(ctx context.Context, c *LogContext, w http.ResponseWriter, r *http
 	return http.StatusOK, nil
 }
 
-// See RFC 6962 Section 4.8. This is mostly used for debug purposes rather than by normal
-// CT clients.
+// See RFC 6962 Section 4.8. This is mostly used for debug purposes rather than
+// by normal CT clients.
+// TODO(pavelkalinnikov): This is actually very useful for computing hash of a
```
Comment (Martin2112) [inline trillian/ctfe/handlers.go]: Not sure the TODO is useful. Just remove the comment about debug? I'll be using it in DNS server as well.

File: trillian/migrillian/core/controller.go [hunk 4]
```diff
@@ -74,50 +100,51 @@ func (c *Controller) Run(ctx context.Context, ctClient *client.LogClient, trClie
...
+	}
+
+	hash, err := merkle.NewLogVerifier(h).VerifiedPrefixHashFromInclusionProof(
+		int64(root.TreeSize), int64(sth.TreeSize),
+		resp.AuditPath, sth.SHA256RootHash[:], leafHash)
+	if err != nil {
+		return err
+	}
+	if !bytes.Equal(hash, root.RootHash) {
+		return errors.New("inconsistent root")
```
Comment (Martin2112) [inline trillian/migrillian/core/controller.go]: Include mismatched hashes to help debug it maybe.

File: trillian/migrillian/core/trillian.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/google/certificate-transparency-go/scanner"
+	"github.com/google/trillian"
+	"github.com/google/trillian/client"
+	"github.com/google/trillian/crypto"
+	"github.com/google/trillian/types"
+)
+
+// TrillianTreeClient is a means of communicating with a Trillian pre-ordered
+// log tree.
+type TrillianTreeClient struct {
```
Comment (Martin2112) [inline trillian/migrillian/core/trillian.go]: Name could be improved. If it's specific to preordered log the name is too generic. PreorderedLogClient? Though it's not directly analagous to TrillianLogClient, which means perhaps Client shouldn't be there.

File: trillian/migrillian/core/trillian.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	cli    trillian.TrillianLogClient
+	verif  *client.LogVerifier
+	tree   *trillian.Tree
+	prefix string // TODO(pavelkalinnikov): Get rid of this.
+}
+
+// NewTrillianTreeClient creates and initializes a pre-ordered log client.
+func NewTrillianTreeClient(ctx context.Context,
```
Comment (Martin2112) [inline trillian/migrillian/core/trillian.go]: Might be more flexible and probably easier to test if the Tree was passed in avoiding need for admin server here.

File: trillian/migrillian/main.go [hunk 7]
```diff
@@ -82,6 +68,11 @@ func main() {
 		ResponseHeaderTimeout: 30 * time.Second,
 		ExpectContinueTimeout: 1 * time.Second,
 	}
+	// TODO(pavelkalinnikov): Share this between multiple CT clients.
+	httpClient := &http.Client{
+		Timeout:   10 * time.Second,
```
Comment (Martin2112) [inline trillian/migrillian/main.go]: Might need to be more configurable as well. Can be done later.

File: trillian/migrillian/core/trillian.go (hunk 5 above)
Comment (pavelkalinnikov) [reply trillian/migrillian/core/trillian.go]: Couldn't come up with anything better than `PreorderedLogClient`. Is that okay? Other ideas were: `(PO|Preordered)Log(Client|Instance|Gate|Shim|Thing)`.

**Comment by pavelkalinnikov** [comment]
ping

**Comment by Martin2112** [comment]
Naming things is hard. Maybe that file should be renamed to client.go? The generic name of the package "core' makes it harder to shorten the name so PreorderedLogClient will be probably be OK for the moment.

**Pull request merged**